src/util/heap/space_descriptor.rs

```diff
@@ -36,12 +32,12 @@ impl SpaceDescriptor {
     pub const UNINITIALIZED: Self = SpaceDescriptor(0);
 
     pub fn create_descriptor_from_heap_range(start: Address, end: Address) -> SpaceDescriptor {
-        let top = end == vm_layout_constants::HEAP_END;
+        let top = end == VM_LAYOUT_CONSTANTS.heap_end;
         if cfg!(target_pointer_width = "64") {
```

Comment:
Although this is not documented well, my understanding is that `create_descriptor_from_heap_range` can only be called on contiguous spaces (for both 32-bit and 64-bit architecture). The function implements two descriptor calculation policy, based on the architecture pointer width, i.e. `cfg!(target_pointer_width = "64")`. This is not depend on the VM layout constants.

Reply:
`create_descriptor_from_heap_range` has only one caller, and the caller only call this for contiguous spaces.

Reply:
The current `master` code assumes "64 bits === 64 bits VM layout === contiguous space". But this assumption is no longer true with the changes in the PR.

In the case where we use the compressed pointer layout on a 64 bits machine, `space_shift_64` is not set for compressed pointer layout, and should not be used. But the code here still calculates the descriptor from `space_shift_64` which would be incorrect. I think we should use the code from L51 for the compressed pointer layout, right? This is the reason I suggested that it should depends on the layout rather than the pointer width.

Reply:
Fixed. Changed this to check on vm_layout instead.